protocol/finalizeblock/event_stager.go

```diff
@@ -0,0 +1,81 @@
+package finalizeblock
+
+import (
+	"encoding/binary"
+
+	"cosmossdk.io/store/prefix"
+	storetypes "cosmossdk.io/store/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
+	ante_types "github.com/dydxprotocol/v4-chain/protocol/app/ante/types"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+)
+
+// EventStager supports staging and retrieval of events (of type T) from FinalizeBlock.
+type EventStager[T proto.Message] struct {
+	transientStoreKey    storetypes.StoreKey
+	cdc                  codec.BinaryCodec
+	stagedEventCountKey  string
+	stagedEventKeyPrefix string
+}
+
+// NewEventStager creates a new EventStager.
+func NewEventStager[T proto.Message](
+	transientStoreKey storetypes.StoreKey,
+	cdc codec.BinaryCodec,
+	stagedEventCountKey string,
+	stagedEventKeyPrefix string,
+) EventStager[T] {
+	return EventStager[T]{
+		transientStoreKey:    transientStoreKey,
+		cdc:                  cdc,
+		stagedEventCountKey:  stagedEventCountKey,
+		stagedEventKeyPrefix: stagedEventKeyPrefix,
+	}
+}
+
+// GetStagedFinalizeBlockEventsFromStore retrieves all staged events from the store.
+func (s EventStager[T]) GetStagedFinalizeBlockEventsFromStore(
+	store storetypes.KVStore,
```

Comment:
also meganit: take in `ctx` and move the free gas meter here for consistency w setter method